examples/todoapp/bin/server/controllers/todos_controller.dart

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2023, Prime Holding JSC
+// https://www.primeholding.com
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'package:shelf/shelf.dart';
+import 'package:shelf_router/shelf_router.dart';
+
+import '../services/todos_service.dart';
+import '../utils/api_controller.dart';
+import '../utils/server_exceptions.dart';
+
+class TodosController extends ApiController {
+  TodosController(this.todosService);
+
+  final TodosService todosService;
+
+  @override
+  void registerRequests(WrappedRouter router) {
+    router.addRequest(
+      RequestType.GET,
+      '/api/v1/todos',
+      getAllTodosHandler,
+    );
+    router.addRequest(
+      RequestType.POST,
+      '/api/v1/todos',
+      addTodoHandler,
+    );
+    router.addRequest(
+      RequestType.PUT,
+      '/api/v1/todos/<id>',
+      updateTodoByIdHandler,
+    );
+    router.addRequest(
+      RequestType.PATCH,
+      '/api/v1/todos/<id>',
+      updateCompletedByIdHandler,
+    );
+    router.addRequest(
+      RequestType.PATCH,
+      '/api/v1/todos',
+      updateCompletedForAll,
+    );
+    router.addRequest(
+      RequestType.DELETE,
+      '/api/v1/todos/completed',
+      deleteCompletedHandler,
+    );
+    router.addRequest(
+      RequestType.DELETE,
+      '/api/v1/todos/<id>',
+      deleteByIdHandler,
+    );
+    router.addRequest(
+      RequestType.GET,
+      '/api/v1/todos/<id>',
+      fetchTodoByIdHandler,
+    );
+  }
+
+  Future<Response> getAllTodosHandler(Request request) async {
+    final allTodos = todosService.fetchAllTodos();
+    return responseBuilder.buildOK(data: allTodos.toJson());
+  }
+
+  Future<Response> updateCompletedForAll(Request request) async {
+    final params = await request.bodyFromFormData();
+    final completed = params['completed'];
+    if (completed == null) {
+      throw BadRequestException('completed is required');
+    }
+    final todos = todosService.updateCompletedForAll(completed);
+    return responseBuilder.buildOK(data: todos.toJson());
+  }
+
+  Future<Response> updateTodoByIdHandler(Request request) async {
+    final params = await request.bodyFromFormData();
+    final title = params['title'];
+    final description = params['description'];
```

Comment:
@StanevPrime is there some kid of convention where the param names can be extracted as a constant? For example in the model class/file?

Reply:
Yes the convention is to use json_serializable for ... serialisation and deserialisation. In this case we deal with the TodoModel, so we can use TodoModel.fromJson() method. @JvnSlv do you see any issues using this method ?

Reply:
Not at all, It can be done either way, but since there are only 2 params passed (title and optional description), I thought that it was not necessary to pass it as a model

Reply:
I was thinking of exporting parameters name in constants to be centralized and avoid typos.

In Kotlin I usually do something like the code below:
<img width="800" alt="image" src="https://github.com/user-attachments/assets/3de05a9f-7eb6-4b0b-b277-bf23e34e2383">

Reply:
@JvnSlv let's try to use the TodoModel.fromJson() to avoid the overhead proposed by @nikolay-vasilev-prime